islandora-preservica-validation/rdfUpdate.py

```diff
@@ -0,0 +1,118 @@
+###########################################################
+import sys, os, csv, shutil
+from xml.etree import ElementTree as etree
+from xml.etree.ElementTree import Element, SubElement
+import subprocess,fnmatch 
+from pprint import pprint
+
+f_path = os.path.dirname(os.path.realpath(__file__)) 
+curr = os.getcwd()
+user = os.environ['USER'] if os.getenv("USER") is not None else os.environ['USERNAME']
+
+valid_result_f = "./output/valid_result.csv" 
+f_output = "./output/temp_pid.csv"
+extRel_fpath =  "/output/extfiles"  
+update_fpath = "/output/update_extfiles"
+ 
+#1. create a temp pidsFile holding the verified the bitstreams count for the corresponding perservica obj
+##  result: f_output/temp_pid.csv
+def getVerifiedPids(f_pids):
+    header_fields=["PID"]
+    with open (os.path.join( f_path, f_pids ), 'r', newline='') as pids_f:
+        csvreader = csv.DictReader(pids_f)
+        with open(os.path.join(f_path, f_output), 'w', newline='') as temp_f:
+            csvwriter = csv.DictWriter(temp_f, fieldnames=header_fields)
+            for r in csvreader:
+                if (r["bitstreamCount"] and (r["isValid"]=="Y")):
+                    csvwriter.writerow({header_fields[0]:r["PID"]})
+
+#2.Iterate the intake pids and extract the ext-rel file via drush from islandora
+# and export the original rdf files to the designed output directory
+def drushfetchDatastream(): 
+    pidtest_name = os.path.join(f_path, f_output)
+    try:
+        subprocess.check_call (['drush', '--root=/var/www/html/drupal7/', '--user={}'.format(user), \
+    '--uri=http://gamera.library.pitt.edu', 'islandora_datastream_crud_fetch_datastreams', '--dsid=RELS-EXT', \
+    '--pid_file={}'.format(pidtest_name), '--datastreams_directory={}{}'.format(curr, extRel_fpath), '--filename_separator=^', '-y'])
+   
+    except subprocess.CalledProcessError as e: 
+	    print(f"Command failed with return code {e.returncode}")
+
+#3. helper function to update the xml
+def fileProcess(fname, ele_name, ele_val):
+    #updatefiles in updated_file dir
+    curr_file = curr + extRel_fpath + "/" + fname
+
+    #register ns to reserve the original prefix
+    ns_dict =dict([node for _,node in etree.iterparse(curr_file, events=['start-ns'])])
+    #pprint(ns_dict)
+    etree.register_namespace('', 'http://digital.library.pitt.edu/ontology/relations#')
+    etree.register_namespace('fedora', 'info:fedora/fedora-system:def/relations-external#')
+    etree.register_namespace('fedora-model', 'info:fedora/fedora-system:def/model#')
+    etree.register_namespace('islandora', 'http://islandora.ca/ontology/relsext#')
+    etree.register_namespace('rdf', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#')
+
+    curr_tree = etree.parse(curr_file)
+    curr_root =curr_tree.getroot()
+    #add new tag "preservicaChildCount"
+    desc_element = curr_root.find('rdf:Description', ns_dict)
+    newNode =etree.SubElement(desc_element, etree.QName(ns_dict["islandora"], ele_name))
```

Comment:
If this node already exists in the tree, will this add a new one?  We should delete/replace any existing `preservicaChildCount` node.